### ISSUE TITLE
feat: Structured output with Outlines and xgrammar

### ISSUE THREAD

**Issue by GandalfTea** [firstbatchxyz/dnet#45]
- [ ] Allow Outlines and XGrammar to mask the logits before sampling a new token

- [ ] Try to default to XGrammar for much better performance on the hot path
